### PR TITLE
Default to OVNKubernetes NetworkType for E2E, enable API access for production

### PR DESCRIPTION
**Pull request by drewandersonnz**
### Which issue this PR addresses:

ADO [9823641](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9823641/)

### What this PR does / why we need it:

Adds SDNProvider to the API
Sets OVNKubernetes as the default option for E2E
Maintains OpenShiftSDN as the default option for production

### Test plan for issue:

Unit tests exists for API behavior (`make test-go`) 
E2E tests will transition to OVNKubernetes
 
### Is there any documentation that needs to be updated for this PR?

Auto-generated Swagger API client docs are available in PR.

### CODE REVIEW

**Comment by drewandersonnz** [comment]
Looks to be blocked by https://github.com/mjudeikis/installer/pull/2

File: pkg/api/openshiftcluster.go
Comment (mjudeikis) [inline pkg/api/openshiftcluster.go:201]: name we agreed was `SDNProvider`, why changing? NetworkType is ambiguous. 

https://github.com/Azure/ARO-RP/pull/1200/files#diff-32cee431764117c8a02c5eabb82bbf3fa57167d1297f68164cdf154269b8c74aR113

File: pkg/api/v20210131preview/openshiftcluster.go
Comment (mjudeikis) [inline pkg/api/v20210131preview/openshiftcluster.go:130]: This needs to be Custom Type (as per 1200)

**Review by mjudeikis — CHANGES_REQUESTED** [review]
Still needs wiring into Installer and maybe E2E tests. Good start :)

File: pkg/api/openshiftcluster.go
Comment (drewandersonnz) [reply pkg/api/openshiftcluster.go:201]: > name we agreed

where was that discussion held? I don't think I was a part of that discussion? 

`SDNProvider` is not the name used in the installer, why change the name to something new? 

I agree it isn't the best name, but if we want to change it we should open a PR upstream to installer and request a change there first. We should use the same name, and same JSON location, as the installer expects to maintain consistency in APIs and reduce confusion.

I vote to remain with `NetworkType` to continue to match the current installer implementation.

File: pkg/api/v20210131preview/openshiftcluster.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster.go:130]: Sure, I will make this change shortly

**Comment by drewandersonnz** [comment]
> Still needs wiring into Installer and maybe E2E tests 
 
 @mjudeikis what do you mean? 
 
 * installer is already updated, this PR builds clusters with `OVNKubernetes` by default, by passing `OVNKubernetes` as the default parameter to the RP API, and passing that through to installer after validation. 
 
 * E2E tests will build an `OVNKubernetes` cluster by default, are you requesting we double the E2E tests to do one set for `OpenShiftSDN` and another E2E for `OVNKubernetes` ?

File: pkg/api/openshiftcluster.go
Comment (mjudeikis) [reply pkg/api/openshiftcluster.go:201]: This was like 9 months ago in arch meeting. 
Installer API is not a reference for us as its not RestAPI and naming there is scapfolded from MachineAPI, installer team opinions, cloud provided code base and kubernetes object namings with no consistency towards being rest api.
NetworkType is wrong overloaded term, one example bellow (and there is way more Types out there). So when it lands into APi it becomes confusing even more. 
```
LAN(Local Area Network)
PAN(Personal Area Network)
MAN(Metropolitan Area Network)
WAN(Wide Area Network)
```

Comment (mjudeikis) [reply pkg/api/openshiftcluster.go:201]: > I agree it isn't the best name, but if we want to change it we should open a PR upstream to installer and request a change there first.

I dont think its our position to do so and installer team will never accepts this as it will be breaking change for them. Just because somebody named fields wrong, we dont need to blindly follow them.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic.go
Comment (m1kola) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic.go:218]: switch-case sounds more appropriate for this check. Similar to API server visibility, provisioning state. It will be easier to extend it if we need at some point.:

```go
	switch ap.Visibility {
	case VisibilityPublic, VisibilityPrivate:
	default:
		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, path+".visibility", "The provided visibility '%s' is invalid.", ap.Visibility)
	}
```

File: pkg/cluster/generateconfig.go
Comment (m1kola) [inline pkg/cluster/generateconfig.go:116]: Try to create a cluster not via `2021-01-31-preview` api version: I think you wil get an empty string here and cluster creation will probably fail.

**Comment by mjudeikis** [comment]
> installer is already updated, this PR builds clusters with OVNKubernetes by default, by passing OVNKubernetes as the default parameter to the RP API, and passing that through to installer after validation. 
Missed that one. Sorry.

**Review by m1kola — CHANGES_REQUESTED** [review]

File: pkg/api/openshiftcluster.go
Comment (drewandersonnz) [reply pkg/api/openshiftcluster.go]: Changed as requested.

File: pkg/api/v20210131preview/openshiftcluster.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster.go]: Changed as requested.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic.go]: Updated.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic.go]: You're right, I think it may have been a relic. Removed.

File: pkg/cluster/generateconfig.go
Comment (drewandersonnz) [reply pkg/cluster/generateconfig.go]: This should not be the case, as the parameter has been defaulted to "OVNKubernetes" here: https://github.com/Azure/ARO-RP/pull/1571/files#diff-c7dba7bf3908043db4c458af9edd806998129f1ef7f58ee8fa54a312ce93f23cR356 this is not API version dependent.

Comment (m1kola) [reply pkg/cluster/generateconfig.go]: @drewandersonnz the problem is that `pkg/util/cluster/cluster.go`:

1. Is not being used in production clsuter creation. We use this util in `./hack/cluster create` dev tool and in e2e cluster creation.
2. It actually is API version dependant. We populate internal representation (`api.OpenShiftCluster`) in this util (for simplicity, I guess), but then convert it to an external (API version specific) representation and fire creation to a specific API version:
    * If we are not in dev env (not sure where it is used actually), we create using v20200430. In this case the field will be droped.
      https://github.com/Azure/ARO-RP/blob/d70dc5b473e3b63f525d5c8396f685bc08cbf509/pkg/util/cluster/cluster.go#L411-L417
    * If we are in dev env, we create using v20210131preview. This applies to CI and this is why CI reports green on this RP (which is bad and which is why we need better e2e strategy)
      https://github.com/Azure/ARO-RP/blob/d70dc5b473e3b63f525d5c8396f685bc08cbf509/pkg/util/cluster/cluster.go#L396-L402

I recommend you to create a cluster using older version by either:

* Using az CLI
* Or firing an API request with `curl`. Example is [here](https://github.com/Azure/ARO-RP/pull/1569), but you don't need keyvault and disc encrpytion sets. And you need to use older API version.

Comment (m1kola) [reply pkg/cluster/generateconfig.go]: Actually... It might also be the case for the new api version: `sdnProvider` is optional (`omitempty`). So if you use the new api version, but create a cluster without providing SDN provider.

Basically these are the cases you need to take care of:

* Cluster creation with older API versions (where we can't provide `sdnProvider`)
* Cluster creation with the new API version where you omit `sdnProvider (same as removing `SDNProvider: "OVNKubernetes"` from `pkg/util/cluster/cluster.go`).
* Cluster creation with the new API version where you provide `sdnProvider`

So far the PR seem to handle only the last one.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic.go
Comment (mjudeikis) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic.go:217]: Remove?

File: pkg/api/v20210131preview/openshiftcluster_validatestatic.go
Comment (mjudeikis) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic.go:219]: If you look inside `NewCloudError` it is already doign message merging with `message` and `args...` so `Sprintf` here is redundant. 
```
func NewCloudError(statusCode int, code, target, message string, a ...interface{}) *CloudError {
	return &CloudError{
		StatusCode: statusCode,
		CloudErrorBody: &CloudErrorBody{
			Code:    code,
			Message: fmt.Sprintf(message, a...),
			Target:  target,
		},
	}
}
```

File: pkg/api/v20210131preview/openshiftcluster_validatestatic.go
Comment (mjudeikis) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic.go:218]: `The provided SDNProvider '%s' is invalid.", value`
We dont provide enum values in the error messages. This is not good practice to leak API types into error static messages as it makes user side parsing complicated.

**Review by mjudeikis — CHANGES_REQUESTED** [review]
As mentioned by @m1kola this is not yet done as per 3 usecases.
In example if you create this using old API you get this now:
```yaml
ERRO[2021-07-13T08:43:47+01:00]pkg/util/steps/runner.go:34 steps.Run() step [Action github.com/Azure/ARO-RP/pkg/cluster.(*manager).Install.func1] encountered error: networki
ng.networkType: Required value: network provider type required  client_principal_name= client_request_id=01ea3a82-e3ae-11eb-bf99-b46921811e1f component=backend correlation_i
d= request_id=bff09dab-7500-4c53-957f-2a3735a1152d resource_group=v4-westeurope resource_id=/subscriptions/225e02bc-43d0-43d1-a01a-17e584a4ef69/resourcegroups/v4-westeurope/
providers/microsoft.redhatopenshift/openshiftclusters/mjudeikis2 resource_name=mjudeikis2 subscription_id=225e02bc-43d0-43d1-a01a-17e584a4ef69                               
ERRO[2021-07-13T08:43:48+01:00]pkg/backend/openshiftcluster.go:221 backend.(*openShiftClusterBackend).updateAsyncOperation.func1() networking.networkType: Required value: ne
twork provider type required  client_principal_name= client_request_id=01ea3a82-e3ae-11eb-bf99-b46921811e1f component=backend correlation_id= request_id=bff09dab-7500-4c53-9
57f-2a3735a1152d resource_group=v4-westeurope resource_id=/subscriptions/225e02bc-43d0-43d1-a01a-17e584a4ef69/resourcegroups/v4-westeurope/providers/microsoft.redhatopenshif
t/openshiftclusters/mjudeikis2 resource_name=mjudeikis2 subscription_id=225e02bc-43d0-43d1-a01a-17e584a4ef69  
```

File: pkg/cluster/generateconfig.go
Comment (drewandersonnz) [reply pkg/cluster/generateconfig.go]: Sorry @m1kola you are right. This was my misunderstanding in the way the object was being handled, the generateconfig.go now handles this case.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic.go]: Updated.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic.go]: Sorry, I didn't see that. Thanks for pointing it out!

File: pkg/api/v20210131preview/openshiftcluster_validatestatic.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic.go]: Removed. /facepalm

File: pkg/cluster/generateconfig.go
Comment (m1kola) [inline pkg/cluster/generateconfig.go:104]: In `github.com/Azure/ARO-RP/pkg/cluster` we should not import anything from API versions: these values from external representation. Here we should work with internal representation only.

I imagine that we need something like:

```go
SDNProvider := "OVNKubernetes" // or "OpenShiftSDN", if we still want to have it be defaul
if m.doc.OpenShiftCluster.Properties.NetworkProfile.SDNProvider != "" {
    SDNProvider = m.doc.OpenShiftCluster.Properties.NetworkProfile.SDNProvider
}

// And then use SDNProvider to instantiate an `InstallConfig`.
```

Comment (m1kola) [reply pkg/cluster/generateconfig.go:104]: Alternatively, we need a generic way to set defaults. But given that at the moment we have only one use case where we need to set defaults, I'm not sure that it is worth doing something more complicated at this point.

**Comment by drewandersonnz** [comment]
Rebased.

File: pkg/cluster/generateconfig.go
Comment (drewandersonnz) [reply pkg/cluster/generateconfig.go]: Ack, I've taken your example.

File: pkg/cluster/generateconfig.go
Comment (drewandersonnz) [reply pkg/cluster/generateconfig.go]: Created default here: https://github.com/Azure/ARO-RP/pull/1571#pullrequestreview-707279749

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (m1kola) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go:477]: Can we have an extra case for when `SDNProvider` is empty?

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (m1kola) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go:485]: We need different names for cases.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (m1kola) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go:466]: I think lets put it into `TestOpenShiftClusterStaticValidateNetworkProfile`.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (m1kola) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go:495]: Take a look at tests where we have `commonTests`, `createTests` and `updateTests`: we want to reduce copy-paste, if possible.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (m1kola) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go:501]: `"Changing property '*' is not allowed"` tests need to go `TestOpenShiftClusterStaticValidateDelta` test which tests immutability of the fields.

File: pkg/cluster/generateconfig.go
Comment (m1kola) [inline pkg/cluster/generateconfig.go:103]: I think we can drop the comment: it should be clear that `"OVNKubernetes"` is default and it is possible to override it.

**Review by m1kola — CHANGES_REQUESTED** [review]
Looks good, but we need to tidy up the tests before we merge.

File: pkg/api/v20210131preview/openshiftcluster_convert.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_convert.go]: Added.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go]: Added.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go]: Names changed.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go]: Tests moved.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go]: Moved to commontests

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go:487]: Moved.

File: pkg/cluster/generateconfig.go
Comment (drewandersonnz) [reply pkg/cluster/generateconfig.go]: Dropped.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (m1kola) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go:481]: I think we should drop this test: I believe it does not test anything.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (m1kola) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go:81]: Use the constant here.

File: pkg/cluster/generateconfig.go
Comment (m1kola) [inline pkg/cluster/generateconfig.go:106]: I think in `pkg/api/openshiftcluster.go` we need to define constants for each SDN provider, simialr to what we do in the external representation:

```go
// SDNProvider constants.
type SDNProvider string

const (
	SDNProviderOVNKubernetes SDNProvider = "OVNKubernetes"
	SDNProviderOpenShiftSDN  SDNProvider = "OpenShiftSDN"
)
```

And we should replace hardcoded `"OVNKubernetes"` here with `api.SDNProviderOVNKubernetes` here.

File: pkg/api/admin/openshiftcluster_validatestatic_test.go
Comment (m1kola) [inline pkg/api/admin/openshiftcluster_validatestatic_test.go:321]: Use constant here.

File: pkg/api/openshiftcluster.go
Comment (m1kola) [inline pkg/api/openshiftcluster.go:196]: We need to define constants for possible values, I think. Similar to external representations.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (m1kola) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go:418]: Constant

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (m1kola) [inline pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go:810]: Constant

File: pkg/util/cluster/cluster.go
Comment (m1kola) [inline pkg/util/cluster/cluster.go:356]: Constant from internal representation: `api.SDNProviderOVNKubernetes`

**Review by m1kola — CHANGES_REQUESTED** [review]
I just realised that we are not defining constatns for the internal representnation in `pkg/api/openshiftcluster.go`.

I think the PR looks very clean and I hope it is the last round of review from me, becuase everything else looks good.

**Comment by drewandersonnz** [comment]
Rebased (again)

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go]: https://github.com/Azure/ARO-RP/pull/1571/commits/7cf26fefa84635caced03ddb15651f0635027f4f#diff-d2a80e56830699a55c16cd21471ad6f85b8061d822f37b247f4b5021f337dfe0L476-L481 

Removed.

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go]: https://github.com/Azure/ARO-RP/pull/1571/commits/7cf26fefa84635caced03ddb15651f0635027f4f#diff-d2a80e56830699a55c16cd21471ad6f85b8061d822f37b247f4b5021f337dfe0R81 

Constant

File: pkg/cluster/generateconfig.go
Comment (drewandersonnz) [reply pkg/cluster/generateconfig.go]: https://github.com/Azure/ARO-RP/pull/1571/commits/7cf26fefa84635caced03ddb15651f0635027f4f#diff-f9a5398ee035bbae6c18fc8468081bbbd9bab4d81504809a5589541ca2171aa7R198-R202

Constants

Comment (drewandersonnz) [reply pkg/cluster/generateconfig.go]: https://github.com/Azure/ARO-RP/pull/1571/commits/7cf26fefa84635caced03ddb15651f0635027f4f#diff-2e16e231ac7cc84006ec5318e10e543d7d827adc2f36f3b7483d90665d6274abR103 

Used constant in default.

File: pkg/api/admin/openshiftcluster_validatestatic_test.go
Comment (drewandersonnz) [reply pkg/api/admin/openshiftcluster_validatestatic_test.go]: https://github.com/Azure/ARO-RP/pull/1571/commits/7cf26fefa84635caced03ddb15651f0635027f4f#diff-ae226be87a9c198e106b41e4ee1eb4bfe9b8ce7f83d258c1865182f964b5612eR321 

Constant

File: pkg/api/openshiftcluster.go
Comment (drewandersonnz) [reply pkg/api/openshiftcluster.go:196]: https://github.com/Azure/ARO-RP/pull/1571/commits/7cf26fefa84635caced03ddb15651f0635027f4f#diff-f9a5398ee035bbae6c18fc8468081bbbd9bab4d81504809a5589541ca2171aa7R198-R202

Defined

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go]: https://github.com/Azure/ARO-RP/pull/1571/commits/7cf26fefa84635caced03ddb15651f0635027f4f#diff-d2a80e56830699a55c16cd21471ad6f85b8061d822f37b247f4b5021f337dfe0R418 

Constant

File: pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go
Comment (drewandersonnz) [reply pkg/api/v20210131preview/openshiftcluster_validatestatic_test.go]: https://github.com/Azure/ARO-RP/pull/1571/commits/7cf26fefa84635caced03ddb15651f0635027f4f#diff-d2a80e56830699a55c16cd21471ad6f85b8061d822f37b247f4b5021f337dfe0R839 

Constant.

File: pkg/util/cluster/cluster.go
Comment (drewandersonnz) [reply pkg/util/cluster/cluster.go]: https://github.com/Azure/ARO-RP/pull/1571/commits/7cf26fefa84635caced03ddb15651f0635027f4f#diff-c7dba7bf3908043db4c458af9edd806998129f1ef7f58ee8fa54a312ce93f23cR356 

Constant.

**Review by m1kola — APPROVED** [review]
Looks good to me. Let's wait for e2e

**Comment by m1kola** [comment]
E2E failed on vm redeployment. It is probably a known flake. Restarted.

**Review by mjudeikis — APPROVED** [review]

**Pull request merged**
